lightning/restore/restore.go

```diff
@@ -304,48 +305,207 @@ outside:
+	tblName  string
+	stmtType schemaStmtType
+	sql      string
+}
+
+type restoreSchemaWorker struct {
+	ctx   context.Context
+	quit  context.CancelFunc
+	jobCh chan *schemaJob
+	errCh chan error
+	wg    sync.WaitGroup
+	glue  glue.Glue
+	store storage.ExternalStorage
+}
+
+func (worker *restoreSchemaWorker) makeJobs(dbMetas []*mydump.MDDatabaseMeta) {
+	select {
+	case <-worker.ctx.Done():
+		return
+	default:
+		// 1. restore databases, execute statements concurrency
+		for _, dbMeta := range dbMetas {
+			restoreSchemaJob := &schemaJob{
+				dbName: dbMeta.Name,
+				stmts:  make([]*schemaStmt, 0, 1),
+			}
+			worker.wg.Add(1)
+			restoreSchemaJob.stmts = append(restoreSchemaJob.stmts, &schemaStmt{
+				stmtType: schemaCreateDatabase,
+				sql:      createDatabaseIfNotExistStmt(dbMeta.Name),
+			})
+			worker.jobCh <- restoreSchemaJob
+		}
+		worker.wg.Wait()
+		// 2. restore tables, execute statements concurrency
+		for _, dbMeta := range dbMetas {
 			for _, tblMeta := range dbMeta.Tables {
-				tablesSchema[tblMeta.Name] = tblMeta.GetSchema(ctx, rc.store)
+				sql := tblMeta.GetSchema(worker.ctx, worker.store)
+				if sql != "" {
+					stmts, err := createTableIfNotExistsStmt(worker.glue.GetParser(), sql, dbMeta.Name, tblMeta.Name)
+					if err != nil {
+						worker.throw(err)
+						return
+					}
+					restoreSchemaJob := &schemaJob{
+						dbName: dbMeta.Name,
+						stmts:  make([]*schemaStmt, 0, len(stmts)),
+					}
+					for _, sql := range stmts {
+						restoreSchemaJob.stmts = append(restoreSchemaJob.stmts, &schemaStmt{
+							tblName:  tblMeta.Name,
+							stmtType: schemaCreateTable,
+							sql:      sql,
+						})
+					}
+					worker.wg.Add(1)
+					worker.jobCh <- restoreSchemaJob
+				}
 			}
-			err := InitSchema(ctx, rc.tidbGlue, dbMeta.Name, tablesSchema)
-
-			task.End(zap.ErrorLevel, err)
-			if err != nil {
-				return errors.Annotatef(err, "restore table schema %s failed", dbMeta.Name)
+		}
+		worker.wg.Wait()
+		// 3. restore views. Since views can cross database we must restore views after all table schemas are restored.
+		for _, dbMeta := range dbMetas {
+			restoreSchemaJob := &schemaJob{
+				dbName: dbMeta.Name,
+				stmts:  make([]*schemaStmt, 0), // sadly, we can't pre-alloc precise value of cap
+			}
+			for _, viewMeta := range dbMeta.Views {
+				sql := viewMeta.GetSchema(worker.ctx, worker.store)
+				if sql != "" {
+					stmts, err := createTableIfNotExistsStmt(worker.glue.GetParser(), sql, dbMeta.Name, viewMeta.Name)
+					if err != nil {
+						worker.throw(err)
+						return
+					}
+					for _, sql := range stmts {
+						restoreSchemaJob.stmts = append(restoreSchemaJob.stmts, &schemaStmt{
+							tblName:  viewMeta.Name,
+							stmtType: schemaCreateView,
+							sql:      sql,
+						})
+					}
+				}
 			}
+			worker.wg.Add(1)
+			worker.jobCh <- restoreSchemaJob
+			// we don't support restore views concurrency, cauz it maybe will raise a error
+			worker.wg.Wait()
 		}
+		worker.quit()
```

Comment:
I think replace this quit with `close(worker.jobChan)` is enough, since all tasks has finished when the program runs here

Reply:
Agree